src/submodule.c

```diff
@@ -149,6 +149,48 @@ static int find_by_path(const git_config_entry *entry, void *payload)
 	return 0;
 }
 
+/*
+ * Checks to see if the submodule shares its name with a file or directory that
+ * already exists on the index. If so, the submodule cannot be added.
+ */
+static int is_path_occupied(bool *occupied, git_repository *repo, const char *path)
+{
+	int error = 0;
+	git_index *index;
+	git_buf dir = GIT_BUF_INIT;
+	*occupied = false;
+
+	if ((error = git_repository_index__weakptr(&index, repo)) < 0)
+		goto out;
+
+	if ((error = git_index_find(NULL, index, path)) == 0) {
```

Comment:
I'd say it'd be a bit less fragile if the check was `if ((error = git_index_find(NULL, index, path)) != GIT_ENOTFOUND)`. Otherwise, we might just ignore all errors by `git_index_find` (even though it currently does not return any error different than `GIT_ENOTFOUND`.